src/frontier_detection/src/map.cpp

```diff
@@ -0,0 +1,278 @@
+
+    mpRawPointCloud.data.resize(mpRawPointCloud.row_step * mpRawPointCloud.height);
+	mpPointCloudSubscriber = this->create_subscription<MapMsg>(
+			"/orbslam3_mono_node/map_points",
+			10,
+			std::bind(&OctoMapNode::AddNewPointCloudCallback, this, std::placeholders::_1));
+	mpMapPublisher = this->create_publisher<octomap_msgs::msg::Octomap>("~/octomap", 10);
+	mpPointCloudPublisher = this->create_publisher<sensor_msgs::msg::PointCloud2>("~/point_cloud", 10);
+	mpMapPublisherTimer = this->create_wall_timer(
+		std::chrono::milliseconds(100), std::bind(&OctoMapNode::PublishMap, this)
+	);
+	mpThrMsgQueueProcess = new std::thread(&OctoMapNode::UpdateMap, this);	
+}
+
+void OctoMapNode::UpdateMap(){
+	MapMsg::SharedPtr msg;
+	while(true){
+		{
+			std::unique_lock<std::mutex> lock(mpMtxMsgQueue);
+			if(mpMsgQueue.empty()){
+				continue;
+			}
+			else{
+				msg = mpMsgQueue.front();	
+				mpMsgQueue.pop_front();
+			}
+	
+		}
+		octomap::Pointcloud cloud;
+		octomap::point3d point;
+		octomap::pointCloud2ToOctomap(msg->cloud, cloud);
+		// ConvertPointCloud2ToOctomap(msg->cloud, cloud);
+		CreatePoint3dFromTransform(msg->pose, point);
+		mpGlobalMap->UpdateMap(cloud, point);
+	}
+}
+
+void OctoMapNode::UpdateRawPointCloud(sensor_msgs::msg::PointCloud2& cloud) {
+
+    // Verify that both point clouds have the same fields
+    if (cloud.fields != mpRawPointCloud.fields) {
+        RCLCPP_ERROR(this->get_logger(), "PointCloud2 messages have different fields and cannot be combined.");
+		return;
+    }
+
+    // Copy header from the first cloud (modify as needed if frame alignment is required)
+    mpRawPointCloud.header = cloud.header;
+
+    // Set the output point cloud properties
+    mpRawPointCloud.height = 1; // Unordered point cloud
+    mpRawPointCloud.is_bigendian = cloud.is_bigendian;
+    mpRawPointCloud.is_dense = cloud.is_dense;
+    mpRawPointCloud.fields = cloud.fields;
+    mpRawPointCloud.point_step = cloud.point_step;
+    mpRawPointCloud.row_step = 0; // Will be updated later
+
+    // Combine point data
+    mpRawPointCloud.data.reserve(cloud.data.size() + mpRawPointCloud.data.size());
+    mpRawPointCloud.data.insert(mpRawPointCloud.data.end(), cloud.data.begin(), cloud.data.end());
+
+    // Update the total number of points
+    mpRawPointCloud.width = (cloud.width * cloud.height) + (mpRawPointCloud.width * mpRawPointCloud.height);
+    mpRawPointCloud.row_step = mpRawPointCloud.width * mpRawPointCloud.point_step;
+}
+
+void OctoMapNode::AddNewPointCloudCallback(const MapMsg::SharedPtr msg){
+	RCLCPP_DEBUG(this->get_logger(), "Adding point cloud to queue for processing");
+	std::unique_lock<std::mutex> lock(mpMtxMsgQueue);
+	mpMsgQueue.push_back(msg);
+	UpdateRawPointCloud(msg->cloud);
+	sensor_msgs::msg::PointCloud2 cloud = mpRawPointCloud;
+	mpPointCloudPublisher->publish(cloud);
+}
+
+void OctoMapNode::ConvertPointCloud2ToOctomap(const sensor_msgs::msg::PointCloud2 &cloud_msg, octomap::Pointcloud &octomap_cloud) {
```

Comment:
why this separate function?

Reply:
Just for general organisation. Other than that no reason.

Reply:
I prefer writing shorter functions each performing a specific function whose name properly describes what is happening. that is why almost all my functions are here are kind of short.

Reply:
cool